onnxruntime/core/framework/allocation_planner.cc

```diff
@@ -581,15 +581,32 @@ class PlannerImpl {
           AllocPlan(current).alloc_kind = AllocKind::kAllocateOutput;
 
           // hacky perf optimization to not copy a pre-existing value to an output if this is a Loop subgraph.
-          // ideally this is temporary, and a future ONNX change to allow empty variadic inputs means we don't
-          // have converted models that unnecessarily add loop state variables. if the value is just being
-          // passed through an implicit input should be used instead.
+          //
+          // this generally occurred with a loop state variable that was only being provided due to ONNX not
+          // supporting empty variadic inputs (a dummy loop state variable was required to make it happy).
+          // ONNX now supports empty variadic inputs so a new model shouldn't need this to avoid poor performance.
+          // problematic older models still exist and the optimization remains for now.
+          //
+          // it has also been triggered when a value from outer scope has being explicitly passed in
```

Comment:
Minor nit: "It is also triggered when a value from outer scope is explicitly passed in" ?? This is still true and desired, right?

Reply:
It will still be triggered and desired, however there's no reason to use a loop state variable in this case as using the outer scope value directly provides the same behavior.